### PR TITLE
TypeFunction.callMatch: Improve structure and control flow

### PR DESCRIPTION
**Pull request by Geod24**
Just a refactoring, but a needed one.

### CODE REVIEW

File: compiler/src/dmd/mtype.d [hunk 1]
```diff
@@ -7343,3 +7192,157 @@ private extern(D) bool isCopyConstructorCallable (StructDeclaration argStruct,
...
+                return MATCH.nomatch;
+            m = MATCH.exact;
+        }
+
+        import dmd.dcast : cimplicitConvTo;
+        m = (sc && sc.flags & SCOPE.Cfile) ? arg.cimplicitConvTo(tprm) : arg.implicitConvTo(tprm);
+    }
+
+    // Non-lvalues do not match ref or out parameters
+    if (p.isReference())
```
Comment (thewilsonator) [inline compiler/src/dmd/mtype.d:7252]: could use an early return here.

**Review by thewilsonator — APPROVED** [review]
Otherwise looks great!

File: compiler/src/dmd/mtype.d (hunk 1 above)
Comment (Geod24) [reply compiler/src/dmd/mtype.d:7252]: I don't think an early return makes the code clearer here. If anything, the reference code could be separated in its own function, which IMO would be better. But keep in mind I'm doing those refactors to make it easier to implement named parameters.

**Comment by Geod24** [comment]
Reduced the scope of the PR to some smaller improvements, while I'm trying to untangle the rest.

**Comment by thewilsonator** [comment]
I'll leave merging to you.

**Pull request merged**